[Bug] [Cron] Parse corn expression error

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

Crontab check and schedule use two crontab classes but will throw `org.apache.dolphinscheduler.service.exceptions.CronParseException: Parse corn expression:` exception.


### What you expected to happen

Runs expect.

### How to reproduce

1. Config crontab use tools.
![image](https://user-images.githubusercontent.com/20885366/229078982-605c301f-a834-4986-906b-658e19c9d54c.png)
2. Click `Execute time` and get an error.
<img width="507" alt="image" src="https://user-images.githubusercontent.com/20885366/229079094-99f70d31-7242-4e74-ab00-57a28fbb4e38.png">
3. But can submit success.
<img width="354" alt="image" src="https://user-images.githubusercontent.com/20885366/229080281-d917aef7-5380-42dc-bede-31cb7941041a.png">
4. Runs error.

```
[ERROR] org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable:[696] - [WorkflowInstance-98287][TaskInstance-0] - Start workflow error
org.apache.dolphinscheduler.service.exceptions.CronParseException: Parse corn expression: [0 0 13/0 * * ? *] error
at org.apache.dolphinscheduler.service.cron.CronUtils.parse2Cron(CronUtils.java:84)
at org.apache.dolphinscheduler.service.cron.CronUtils.getFireDateList(CronUtils.java:131)
at org.apache.dolphinscheduler.service.cron.CronUtils.getSelfFireDateList(CronUtils.java:228)
at org.apache.dolphinscheduler.service.cron.CronUtils.getSelfFireDateList(CronUtils.java:198)
at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.initTaskQueue(WorkflowExecuteRunnable.java:887)
at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.call(WorkflowExecuteRunnable.java:685)
at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1604)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.IllegalArgumentException: Failed to parse cron expression. Period 0 not in range [0, 23]
at com.cronutils.parser.CronParser.parse(CronParser.java:131)
at org.apache.dolphinscheduler.service.cron.CronUtils.parse2Cron(CronUtils.java:82)
... 9 common frames omitted
Caused by: java.lang.IllegalArgumentException: Period 0 not in range [0, 23]
```

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java
+++ b/dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java
@@ -122,6 +122,7 @@ import org.apache.dolphinscheduler.plugin.task.api.enums.TaskTimeoutStrategy;
 import org.apache.dolphinscheduler.plugin.task.api.model.Property;
 import org.apache.dolphinscheduler.plugin.task.api.parameters.ParametersNode;
 import org.apache.dolphinscheduler.plugin.task.api.parameters.SqlParameters;
+import org.apache.dolphinscheduler.service.cron.CronUtils;
 import org.apache.dolphinscheduler.service.model.TaskNode;
 import org.apache.dolphinscheduler.service.process.ProcessService;
 
@@ -2607,7 +2608,7 @@ public class ProcessDefinitionServiceImpl extends BaseServiceImpl implements Pro
             putMsg(result, Status.SCHEDULE_START_TIME_END_TIME_SAME);
             return result;
         }
-        if (!org.quartz.CronExpression.isValidExpression(scheduleObj.getCrontab())) {
+        if (!CronUtils.isValidExpression(scheduleObj.getCrontab())) {
             log.error("CronExpression verify failure, cron:{}.", scheduleObj.getCrontab());
             putMsg(result, Status.REQUEST_PARAMS_NOT_VALID_ERROR, scheduleObj.getCrontab());
             return result;

--- a/dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/SchedulerServiceImpl.java
+++ b/dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/SchedulerServiceImpl.java
@@ -75,7 +75,6 @@ import java.util.stream.Collectors;
 import lombok.NonNull;
 import lombok.extern.slf4j.Slf4j;
 
-import org.quartz.CronExpression;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
@@ -182,7 +181,7 @@ public class SchedulerServiceImpl extends BaseServiceImpl implements SchedulerSe
 
         scheduleObj.setStartTime(scheduleParam.getStartTime());
         scheduleObj.setEndTime(scheduleParam.getEndTime());
-        if (!org.quartz.CronExpression.isValidExpression(scheduleParam.getCrontab())) {
+        if (!CronUtils.isValidExpression(scheduleParam.getCrontab())) {
             log.error("Schedule crontab verify failure, crontab:{}.", scheduleParam.getCrontab());
             putMsg(result, Status.REQUEST_PARAMS_NOT_VALID_ERROR, scheduleParam.getCrontab());
             return result;
@@ -238,7 +237,7 @@ public class SchedulerServiceImpl extends BaseServiceImpl implements SchedulerSe
         if (scheduleParam.getStartTime().getTime() > scheduleParam.getEndTime().getTime()) {
             throw new ServiceException(Status.START_TIME_BIGGER_THAN_END_TIME_ERROR);
         }
-        if (!CronExpression.isValidExpression(scheduleParam.getCrontab())) {
+        if (!CronUtils.isValidExpression(scheduleParam.getCrontab())) {
             throw new ServiceException(Status.SCHEDULE_CRON_CHECK_FAILED, scheduleParam.getCrontab());
         }
     }
@@ -828,7 +827,7 @@ public class SchedulerServiceImpl extends BaseServiceImpl implements SchedulerSe
 
             schedule.setStartTime(scheduleParam.getStartTime());
             schedule.setEndTime(scheduleParam.getEndTime());
-            if (!org.quartz.CronExpression.isValidExpression(scheduleParam.getCrontab())) {
+            if (!CronUtils.isValidExpression(scheduleParam.getCrontab())) {
                 log.error("Schedule crontab verify failure, crontab:{}.", scheduleParam.getCrontab());
                 putMsg(result, Status.SCHEDULE_CRON_CHECK_FAILED, scheduleParam.getCrontab());
                 return;

--- a/dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/SchedulerServiceTest.java
+++ b/dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/SchedulerServiceTest.java
@@ -243,7 +243,7 @@ public class SchedulerServiceTest extends BaseServiceTestTool {
         Assertions.assertEquals(Status.SCHEDULE_CRON_CHECK_FAILED.getCode(), ((ServiceException) exception).getCode());
 
         // error schedule crontab
-        String badCrontab2 = "0 0 123 * * ? *";
+        String badCrontab2 = "0 0 13/0 * * ? *";
         scheduleCreateRequest.setStartTime(startTime);
         scheduleCreateRequest.setCrontab(badCrontab2);
         exception = Assertions.assertThrows(ServiceException.class,

--- a/dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/SchedulerServiceTest.java
+++ b/dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/SchedulerServiceTest.java
@@ -242,6 +242,14 @@ public class SchedulerServiceTest extends BaseServiceTestTool {
                 () -> schedulerService.createSchedulesV2(user, scheduleCreateRequest));
         Assertions.assertEquals(Status.SCHEDULE_CRON_CHECK_FAILED.getCode(), ((ServiceException) exception).getCode());
 
+        // error schedule crontab
+        String badCrontab2 = "0 0 123 * * ? *";
+        scheduleCreateRequest.setStartTime(startTime);
+        scheduleCreateRequest.setCrontab(badCrontab2);
+        exception = Assertions.assertThrows(ServiceException.class,
+                () -> schedulerService.createSchedulesV2(user, scheduleCreateRequest));
+        Assertions.assertEquals(Status.SCHEDULE_CRON_CHECK_FAILED.getCode(), ((ServiceException) exception).getCode());
+
         // error create error
         scheduleCreateRequest.setCrontab(crontab);
         Mockito.when(scheduleMapper.insert(isA(Schedule.class))).thenReturn(0);

--- a/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/cron/CronUtils.java
+++ b/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/cron/CronUtils.java
@@ -85,6 +85,24 @@ public class CronUtils {
     }
 
     /**
+     * Indicates whether the specified cron expression can be parsed into a
+     * valid cron expression
+     *
+     * @param cronExpression the expression to evaluate
+     * @return a boolean indicating whether the given expression is a valid cron
+     *         expression
+     */
+    public static boolean isValidExpression(String cronExpression) {
+        try {
+            parse2Cron(cronExpression);
+        } catch (CronParseException e) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
      * get max cycle
      *
      * @param cron cron

--- a/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/cron/CronUtils.java
+++ b/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/cron/CronUtils.java
@@ -79,7 +79,7 @@ public class CronUtils {
     public static Cron parse2Cron(String cronExpression) throws CronParseException {
         try {
             return QUARTZ_CRON_PARSER.parse(cronExpression);
-        } catch (Exception ex) {
+        } catch (IllegalArgumentException ex) {
             throw new CronParseException(String.format("Parse corn expression: [%s] error", cronExpression), ex);
         }
     }

--- a/dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/cron/CronUtilsTest.java
+++ b/dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/cron/CronUtilsTest.java
@@ -247,7 +247,7 @@ public class CronUtilsTest {
     }
 
     @Test
-    public void testValid(){
+    public void testValid() {
         Assertions.assertFalse(CronUtils.isValidExpression("0 0 13/0 * * ? *"));
         Assertions.assertTrue(CronUtils.isValidExpression("0 0 13-0 * * ? *"));
     }

--- a/dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/cron/CronUtilsTest.java
+++ b/dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/cron/CronUtilsTest.java
@@ -245,4 +245,10 @@ public class CronUtilsTest {
         expirationTime = CronUtils.getExpirationTime(startTime, CycleEnum.YEAR);
         Assertions.assertEquals("2020-02-07 18:30:00", DateUtils.dateToString(expirationTime));
     }
+
+    @Test
+    public void testValid(){
+        Assertions.assertFalse(CronUtils.isValidExpression("0 0 13/0 * * ? *"));
+        Assertions.assertTrue(CronUtils.isValidExpression("0 0 13-0 * * ? *"));
+    }
 }

